File: MarxUp/MarxUp/Camera.m
//
//  Camera.m
//  MarxUp
//
//  Created by Ognyanka Boneva on 1.11.18.
//  Copyright © 2018 Ognyanka Boneva. All rights reserved.
//

#import "Camera.h"
#import "FileManager.h"
#import <AVFoundation/AVFoundation.h>

@interface Camera ()

@property (strong, nonatomic) AVCaptureSession *photoCaptureSession;

@property (strong, nonatomic) AVCaptureDevice *frontCamera;
@property (strong, nonatomic) AVCaptureDevice *backCamera;

@property (strong, nonatomic) AVCaptureDeviceInput *frontCameraInput;
@property (strong, nonatomic) AVCaptureDeviceInput *backCameraInput;

@property (strong, nonatomic) AVCapturePhotoOutput *photoOutput;

@property (assign) AVCaptureDevicePosition currentPosition;

@property (strong, nonatomic) AVCaptureVideoPreviewLayer *previewLayer;

@end

@implementation Camera

+ (instancetype)newCameraWithOutputView:(UIView *)outputView {
    Camera *new = [Camera new];
    new.photoCaptureSession = [AVCaptureSession new];
    [new.photoCaptureSession setSessionPreset:AVCaptureSessionPresetPhoto];
    new.currentPosition = AVCaptureDevicePositionBack;

    new.frontCameraInput = [new defaultFrontCameraInput];
    new.backCameraInput = [new defaultBackCameraInput];
    [new configSessionWithCurrentDeviceInput];

    new.photoOutput = [new defaultOutput];
    [new.photoCaptureSession addOutput:new.photoOutput];

    [new configOutputToView:outputView];
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
        [new.photoCaptureSession startRunning];
        dispatch_async(dispatch_get_main_queue(), ^{
            new.previewLayer.frame = outputView.bounds;
        });
    });
    
    return new;
}

- (AVCaptureDeviceInput *)defaultFrontCameraInput {
    AVCaptureDevice *frontCamera = [AVCaptureDevice defaultDeviceWithDeviceType:AVCaptureDeviceTypeBuiltInWideAngleCamera mediaType:AVMediaTypeVideo position:AVCaptureDevicePositionFront];
    [frontCamera lockForConfiguration:nil];
    if ([frontCamera isFocusModeSupported:AVCaptureFocusModeAutoFocus]) {
            frontCamera.focusMode = AVCaptureFocusModeAutoFocus;
    }
    [frontCamera unlockForConfiguration];
    return [AVCaptureDeviceInput deviceInputWithDevice:frontCamera error:nil];
}

- (AVCaptureDeviceInput *)defaultBackCameraInput {
    AVCaptureDevice *backCamera = [AVCaptureDevice defaultDeviceWithDeviceType:AVCaptureDeviceTypeBuiltInWideAngleCamera mediaType:AVMediaTypeVideo position:AVCaptureDevicePositionBack];
    [backCamera lockForConfiguration:nil];
    if ([backCamera isFocusModeSupported:AVCaptureFocusModeAutoFocus]) {
            backCamera.focusMode = AVCaptureFocusModeAutoFocus;
    }
    [backCamera unlockForConfiguration];
    return [AVCaptureDeviceInput deviceInputWithDevice:backCamera error:nil];
}

- (AVCapturePhotoOutput *)defaultOutput {
    AVCapturePhotoOutput *output = [AVCapturePhotoOutput new];
    [output setHighResolutionCaptureEnabled:YES];
    [output setLivePhotoCaptureEnabled:NO];
    
    return output;
}

- (void)configSessionWithCurrentDeviceInput {
    if (self.currentPosition == EnumCameraFront) {
        [self.photoCaptureSession addInput:self.frontCameraInput];
    }
    else {
        [self.photoCaptureSession addInput:self.backCameraInput];
    }
}

- (void)configOutputToView:(UIView *)view {
    self.previewLayer = [AVCaptureVideoPreviewLayer layerWithSession:self.photoCaptureSession];
    [self.previewLayer setVideoGravity:AVLayerVideoGravityResizeAspect];
    [view.layer addSublayer:self.previewLayer];
}

- (EnumCamera)currentCamera {
    if (self.currentPosition == AVCaptureDevicePositionFront) {
        return EnumCameraFront;
    }
    return EnumCameraBack;
}

- (void)switchCamera {
    if (self.currentPosition == AVCaptureDevicePositionFront) {
        self.currentPosition = AVCaptureDevicePositionBack;
        [self.photoCaptureSession removeInput:self.frontCameraInput];
        [self.photoCaptureSession addInput:self.backCameraInput];
    }
    else {
        self.currentPosition = AVCaptureDevicePositionFront;
        [self.photoCaptureSession removeInput:self.backCameraInput];
        [self.photoCaptureSession addInput:self.frontCameraInput];
    }
}

- (void)takePhoto {
    AVCapturePhotoSettings *photoSettings = [AVCapturePhotoSettings new];
    [photoSettings setHighResolutionPhotoEnabled:YES];
    [photoSettings setFlashMode:AVCaptureFlashModeOff];
    [photoSettings setAutoStillImageStabilizationEnabled:YES];
    [self.photoOutput capturePhotoWithSettings:photoSettings delegate:self];
}

- (void)stopCamera {
    [self.photoCaptureSession stopRunning];
}

- (void)captureOutput:(AVCapturePhotoOutput *)output didFinishProcessingPhoto:(AVCapturePhoto *)photo error:(NSError *)error {
    NSData *imageData = [photo fileDataRepresentation];
    [FileManager saveImage:imageData atImageURL:nil];
}

@end

File: MarxUpUp/MarxUpUp/Camera/Camera.swift
//
//  Camera.swift
//  MarxUpUp
//
//  Created by Ognyanka Boneva on 6.11.18.
//  Copyright © 2018 Ognyanka Boneva. All rights reserved.
//

import UIKit
import AVFoundation

class Camera: NSObject {

    var cameraInput: CameraInput!
    private let captureSession:AVCaptureSession
    private let captureOutput:AVCapturePhotoOutput = {
            let output = AVCapturePhotoOutput()
            output.isHighResolutionCaptureEnabled = true
            output.isLivePhotoCaptureEnabled = false
            return output
    }()
    private var previewLayer: AVCaptureVideoPreviewLayer
    var position = AVCaptureDevice.Position.back
    let device = UIDevice.current
    
    private var currentVideoOrientation: AVCaptureVideoOrientation {
        return AVCaptureVideoOrientation(rawValue: device.orientation.rawValue) ?? .portrait
    }
    
    init(outputView view: UIView) {
        captureSession = AVCaptureSession()
        captureSession.sessionPreset = AVCaptureSession.Preset.photo
        previewLayer = AVCaptureVideoPreviewLayer(session: captureSession)
        super.init()
        
        cameraInput = CameraInput(withPosition: position)
        guard cameraInput.deviceInput != nil else {
            return
        }

        captureSession.addInput(cameraInput.deviceInput!)
        captureSession.addOutput(captureOutput)
        
        
        previewLayer.videoGravity = AVLayerVideoGravity.resizeAspect
        view.layer.addSublayer(previewLayer)
        
        device.beginGeneratingDeviceOrientationNotifications()
        
        DispatchQueue.global().async {
            self.captureSession.startRunning()
            DispatchQueue.main.async {
                self.previewLayer.connection?.videoOrientation = self.currentVideoOrientation
                self.previewLayer.frame = view.bounds
            }
        }
    }
}

//MARK: - AVCapturePhotoCaptureDelegate Methods
extension Camera: AVCapturePhotoCaptureDelegate {
    func photoOutput(_ output: AVCapturePhotoOutput, didFinishProcessingPhoto photo: AVCapturePhoto, error: Error?) {
        guard error == nil else {
            print("Error while capturing photo: " + error!.localizedDescription)
            return
        }
        
        guard let imageData = photo.fileDataRepresentation() else {
            print("Error: fileDataRepresentation for photo failed")
            return
        }
        DatabaseManager().saveImageWithData(imageData)
    }
}

extension Camera: CameraInterface {
    var isSupportedByTheDevice: Bool {
        return cameraInput.deviceInput != nil
    }
    
    func stop() {
        captureSession.stopRunning()
    }
    
    func start() {
        guard cameraInput.deviceInput != nil else {
            return
        }
        
        DispatchQueue.global().async {
            self.captureSession.startRunning()
        }
    }
    
    func takePhoto() {
        guard cameraInput.deviceInput != nil else {
            return
        }
        
        if let connection = captureOutput.connection(with: .video) {
            connection.videoOrientation = currentVideoOrientation
        }
        
        let photoSetings = AVCapturePhotoSettings()
        photoSetings.isHighResolutionPhotoEnabled = true
        photoSetings.flashMode = .off
        photoSetings.isAutoStillImageStabilizationEnabled = true
        
        captureOutput.capturePhoto(with: photoSetings, delegate: self)
    }
    
    func switchPosition() {
        guard cameraInput.deviceInput != nil else {
            return
        }
        captureSession.removeInput(cameraInput.deviceInput!)
        if position == .back {
            position = .front
        }
        else {
            position = .back
        }
        cameraInput.position = position
        
        guard cameraInput.deviceInput != nil else {
            return
        }
        captureSession.addInput(cameraInput.deviceInput!)
    }
    
    func updateOrientation(forView view: UIView) {
        DispatchQueue.main.async {
            self.previewLayer.connection?.videoOrientation = self.currentVideoOrientation
            self.previewLayer.frame = view.bounds
        }
    }
    
}
